### ISSUE TITLE
repository rules: report slow actions

### ISSUE THREAD

**Issue by laszlocsomor** [bazelbuild/bazel#6459]
### Description of the problem / feature request:

Repository rules should report long-running actions executed by `repository_ctx.execute`. This bug is closely related to https://github.com/bazelbuild/bazel/issues/6458, and they might have the same solution.

### Feature requests: what underlying problem are you trying to solve with this feature?

Report progress in real time for slow repo rules, typically for cases when the repo rule runs an external package manager which downloads a lot of data. Bazel gives no visual progress in this case.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

See https://github.com/laszlocsomor/projects/tree/b30533a64caf43f89ff9c5b812a896b65cb4f270/bazel/slow-repo-rule

### What operating system are you running Bazel on?

Windows 10 and Debian Linux

### What's the output of `bazel info release`?

0.18.0

**Comment by aehlig** [comment]
> Report progress in real time for slow repo rules, typically for cases when the repo rule runs an external package manager which downloads a lot of data. Bazel gives no visual progress in this case.

I would say, this is a duplicate of #1289.

Some progress in this direction was made by https://github.com/bazelbuild/bazel/commit/30cd6ab8cdcefef92b9af6d9f9f0cee06ce2f2fe which will be part of 0.20.0.

-- 
Klaus Aehlig
Google Germany GmbH, Erika-Mann-Str. 33, 80636 Muenchen
Registergericht und -nummer: Hamburg, HRB 86891
Sitz der Gesellschaft: Hamburg
Geschaeftsfuehrer: Paul Terence Manicle, Halimah DeLaine Prado

**Comment by laszlocsomor** [comment]
> I would say, this is a duplicate of #1289.

Thanks! Do you have any idea about ETA for that, or for parts of it?

> Some progress in this direction was made by [30cd6ab](https://github.com/bazelbuild/bazel/commit/30cd6ab8cdcefef92b9af6d9f9f0cee06ce2f2fe) which will be part of 0.20.0.

Great to hear that, thanks!

**Comment by aiuto** [comment]
@laszlocsomor Do you want to close as a dup or leave this open?

**Comment by laszlocsomor** [comment]
I'd like to keep it open for per-command progress reporting (i.e. for slow `repository_ctx.execute` calls).

**Comment by irengrig** [comment]
@aehlig can this be closed after https://github.com/bazelbuild/bazel/issues/6427?

**Comment by aehlig** [comment]
> @aehlig can this be closed after #6427?

I would say yes, but over to @laszlocsomor to decide whether this is what he had in mind, or some form of feature request as like #6659.

**Comment by laszlocsomor** [comment]
#6427 is a great step forward. What I had in mind was more: to report "Still running after XX seconds..." status messages similar to action execution, in case the command runs longer than a given limit (say, 5 seconds). That would give the user visual feedback that everything was still working, just taking longer than they might expect.

**Comment by aehlig** [comment]
> What I had in mind was more: to report "Still running after XX seconds..." status messages similar to action execution, in case the command runs longer than a given limit (say, 5 seconds). 

The current (bazel at HEAD as of now) semantics is that we show
- the last progress message, and
- time time since the beginning of the fetch (not since the last progress message change).

This is slightly different than your suggestion, but at least it is showing that bazel is still alive.

**Comment by laszlocsomor** [comment]
Thanks, these are all valuable progress messages! It's great that we have them now.

> This is slightly different than your suggestion, but at least it is showing that bazel is still alive.

Exactly. I'm not saying what we have at HEAD is insufficient, I'm saying it's not what I'm suggesting, so strictly speaking this bug is not resolved. I'm not claiming that my idea is better, or that it must be done at all.

**Comment by meisterT** [comment]
We show those slow actions at least now in the JSON profile.
